torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -0,0 +1,390 @@
+    memcpy(tmp.data(), value.data(), value.size());
+    store_->set(key, tmp);
+  }
+
+  std::vector<char> get(const std::string& key) override {
+    // @pietern was unable to figure out a way to cast the value of
+    // type std::vector<unsigned char> to an std::vector<char>,
+    // so going with a copy here (it's not a hot path so it's fine).
+    auto value = store_->get(key);
+    std::vector<char> tmp(value.size());
+    memcpy(tmp.data(), value.data(), value.size());
+    return tmp;
+  }
+
+  void wait(const std::vector<std::string>& keys) override {
+    store_->wait(keys, Store::kDefaultTimeout);
+  }
+
+  void wait(
+      const std::vector<std::string>& keys,
+      const std::chrono::milliseconds& timeout) override {
+    store_->wait(keys, timeout);
+  }
+
+ protected:
+  std::shared_ptr<::c10d::Store> store_;
+};
+
+template <typename T>
+const ::gloo::ReductionFunction<T>* reductionFunction(const ReduceOp& r) {
+  switch (r) {
+    case ReduceOp::SUM:
+      return ::gloo::ReductionFunction<T>::sum;
+    case ReduceOp::PRODUCT:
+      return ::gloo::ReductionFunction<T>::product;
+    case ReduceOp::MIN:
+      return ::gloo::ReductionFunction<T>::min;
+    case ReduceOp::MAX:
+      return ::gloo::ReductionFunction<T>::max;
+  }
+
+  throw std::runtime_error("Unhandled ReduceOp");
+}
+
+} // namespace
+
+ProcessGroupGloo::WorkGloo::WorkGloo() {
+  completed_ = false;
+}
+
+ProcessGroupGloo::WorkGloo::~WorkGloo() {
+}
+
+bool ProcessGroupGloo::WorkGloo::isCompleted() {
+  std::unique_lock<std::mutex> lock(m_);
+  return completed_;
+}
+
+bool ProcessGroupGloo::WorkGloo::wait() {
+  std::unique_lock<std::mutex> lock(m_);
+  while (!completed_) {
+    cv_.wait(lock);
+  }
+  return !ex_;
+}
+
+const std::exception& ProcessGroupGloo::WorkGloo::exception() const {
+  return *ex_;
+}
+
+void ProcessGroupGloo::WorkGloo::finish() {
+  {
+    std::unique_lock<std::mutex> lock(m_);
+    completed_ = true;
+  }
+  cv_.notify_all();
+}
+
+void ProcessGroupGloo::WorkGloo::finishWithException(
+    const ::gloo::Exception& ex) {
+  {
+    std::unique_lock<std::mutex> lock(m_);
+    completed_ = true;
+    ex_ = std::unique_ptr<::gloo::Exception>(new ::gloo::Exception(ex));
+  }
+  cv_.notify_all();
+}
+
+ProcessGroupGloo::ProcessGroupGloo(
+    const std::shared_ptr<Store>& store,
+    const std::vector<std::shared_ptr<::gloo::transport::Device>>& devices,
+    int rank,
+    int size)
+    : ProcessGroup(rank, size),
+      store_(new GlooStore(store)),
+      devices_(devices) {
+  stop_ = false;
+}
+
+ProcessGroupGloo::~ProcessGroupGloo() {
+  // Require this process group to be explicitly shut down prior to being
+  // destructed. This is the easiest way to guarantee clean shutdown
+  // and avoid blocking/throwing in a destructor.
+}
+
+void ProcessGroupGloo::initialize() {
+  contexts_.clear();
+  for (auto& device : devices_) {
+    auto context = std::shared_ptr<::gloo::rendezvous::Context>(
```

Comment:
Please use make_shared to avoid the memory leak when "new" failed for some reasons.